iwyu.cc

```diff
@@ -1761,6 +1761,32 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     return true;
   }
 
+  // The explicit traversal of 'CXXForRangeStmt' is introduced to avoid problem
```

Comment:
Some notes on this comment:

* `avoid problem` -> 'avoid problems' or 'avoid a problem'
* `the specific providing type is lost` -- is this because it's been desugared by Clang? If so, maybe say "desugared" instead of "lost"
* It's not clear to me from the text what the higher-level "problem" is we're avoiding here? What is the bad thing that happens if we remove this custom traversal?

I checked out the branch and tried removing this code, and could see that:
```
  for (auto i : ref_provided)
    ;
```
started complaining about a full use of `Indirect`.

So I guess the problem is:

* Types in `CXXForRangeStmt` are implicitly desugared
* The default traversal sees these desugared types and reports them
* The actual type in the written code is a sugared alias type
* The underlying desugared type is provided from the alias definition loc (because it's an alias, and there are no author intent hints to the contrary), so it should never be reported in use locations

I guess this leaves me with the question: how does the non-providing ref [here](https://github.com/include-what-you-use/include-what-you-use/blob/master/tests/cxx/ptr_ref_aliases.cc#L76) get covered if we never visit it?

Or is it that the implicit part of the `CXXForRangeStmt` exposes desugared types and the explicit part the sugared ones?

Reply:
> `the specific providing type is lost` -- is this because it's been desugared by Clang? If so, maybe say "desugared" instead of "lost"

I wanted to say: the type is desugared hence the information that it is provided and should not be reported is lost.

> how does the non-providing ref [here](https://github.com/include-what-you-use/include-what-you-use/blob/master/tests/cxx/ptr_ref_aliases.cc#L76) get covered if we never visit it?

It is reported from `VisitCXXForRangeStmt` handler ([you have written it](https://github.com/include-what-you-use/include-what-you-use/commit/2a474258599b22a3022b503975747658d55234da) especially for that). After fixing the issue with forwarding references, one can remove both `Visit...` and `Traverse...` handlers of `CXXForRangeStmt` and have the same output as with the sole `VisitCXXForRangeStmt`. (But it doesn't mean that `VisitCXXForRangeStmt` is not needed anymore, due to the abovementioned reason.)